### ISSUE TITLE
기존 코드 리팩토링 & `RDBeta` 정의 및 학습하기

### ISSUE THREAD

**Issue by eubinecto** [wisdomify/wisdomify#68]
## Why?
- https://github.com/wisdomify/wisdomify/issues/56#issuecomment-923674600

## What?
- pre-training: 이미 학습된 kcBERT를 활용, 속담만을 MLM으로 학습 - 이때, 속담을  단일 토큰으로 취급하면서도, 속담 속 형태소의 임베딩도 활용을 한다.
- fine-tuning: 정의 데이터에 핏. 알고리즘은 `ver_0`와 동일함.

### Canceled
~- [x] `RDZero`, `RDOne` 정의.~
~- [x] `Builder`, `BuilderZero`, `BuilderOne`을 정의.~
~- [x]  `build_wisdom2subwords`, `build_wiskeys`~
~- [x]  `conf.json`에 토크나이저 이름 / path도 추가하기!~

### 리팩토링
- [x]  `RD`, `RDAlpha`, `RDBeta` 정의
- [x]  `TensorBuilder` 및 하위 빌더 정의
- [x]  변경된 코드에 맞게, 테스트 리팩토링 하기
- [x] 수지님, 유라님이 `RDBeta.S_wisdom_figurative` 만 구현하면 바로 테스트 해볼 수 있도록 환경 구성하기
- [x]  그냥 `XBuilder`도  wisdom mask를 출력하도록 하기

### `RDBeta`
- [x]  @Yuuraa @ohsuz  - `S_wisdom_figurative` 정의하고,  테스트 확인하기#68 

### `RDGamma`
- [ ]  @Yuuraa  -  `RDBeta`의 구조를 따라가돼, linear layer 대신 attention을 사용하여 평균값을 구하는 방법을 시도해보기

**Comment by eubinecto** [comment]
## 구조 고민하기
fine-tuning을 하는 부분의 코드가 달라져야한다. 여기도 pre-training을 할때와 정확히 동일한 구조를 가져야 한다.

**Comment by eubinecto** [comment]
## `Builder`, `BuilderZero`, `BuilderOne`을 정의

**Comment by eubinecto** [comment]
## `build_wisdom2subwords`

**Comment by eubinecto** [comment]
`RD` 클래스에  `S_wisdom` 메소드를 추가하고, `RDAlpha` , `RDBeta` 가 이  함수를 구현하도록 만들면 될 것 같다.
어차피 어떤 모델을 만들게 되던, 결국 계산해야하는 것은 `S_wisdom`이기 때문.

**Comment by eubinecto** [comment]
음.. 만들고자 하는 모델을 3개로 정의하자.
1.  `RDAlpha`
  -  S_wisdom =  S_wisdom_literal
  - trained on:  wisdom2def
2. `RDBeta`
  -  S_wisdom =  S_wisdom_literal + S_wisdom_figurative
  - trained on:  wisdom2def
3. `RDGamma`
  - S wisdom = S_wisdom_literal  +  S_wisdom_figurative
  - trained on: wisdom2def

**Comment by eubinecto** [comment]
빌더는 어떻게 해야하지? 앞으로 새로운 모델을 만들때마다...각 모델이 `X`와  `y`를 정의해줘야 할텐데. 예를들어, `RDAlpha`, `RDBeta`는 요구하는  X, y가 동일하지만.. `RDGamma`가 요구하는  X는 다르다.  전자에서는 속담이 반드시 왼쪽에 존재하지만, 후자에서는 속담이 문장 어디에든 위치할 수 있다.

**Comment by eubinecto** [comment]
모든 리팩토링은 전부 진행완료! 이제 @Yuuraa  @ohsuz 님이 이 부분만 채워주시면 됩니다!

https://github.com/wisdomify/wisdomify/blob/1716b4c5a97a4443c466f81a65c557df32a4098d/wisdomify/rds.py#L171-L179

**Comment by eubinecto** [comment]
아, 생각해보니, `RDBeta`를 사용하는 경우, tokenizer도 세이브를 해야한다. load를 하는 경우에는 저장된 tokenizer를 불러와야 한다..!

**Comment by eubinecto** [comment]
그리고 test 로직도 다시 리팩토링이 필요하다.. 여기에 더해서 infer &  eval 잘 동작하는지도 테스트에 넣고 싶은데, 어떻게 안될려나?

**Comment by eubinecto** [comment]
`RDBeta`를 위한 테스트도 추가완료:

https://github.com/wisdomify/wisdomify/blob/43e1661630e30ff0d4fbf46221e743d25d1bf7ea/wisdomify/tests/test_models.py#L123-L141

@ohsuz @Yuuraa `RDBeta`의  `S_wisdom_figurative` 구현을 완료한 후, `expectedFailure` 데코레이터를 붙여놓은 테스트의 데코레이터를 제거하고 테스트를 실행해주세요. 제대로된 차원을 출력하는지 간단히 테스트해보는 로직입니다. 데코레이터 제거 이후에도 테스트가 무사히 패스한다면 제게 말해주세요!

**Comment by eubinecto** [comment]
수정 - `S_wisdom_figurative`의 입력을 `H_all`로 정의하는편이, 모델링을 하기에 더 편할 것. cls랑 k토큰 외에 다른 토큰의 히든벡터도 사용하고 싶을 때가 있을테니?
```python3
    def S_wisdom(self, H_all: torch.Tensor) -> torch.Tensor:
        H_k = H_all[:, 1: self.hparams['k'] + 1]  # (N, L, H) -> (N, K, H)
        S_wisdom = self.S_wisdom_literal(H_k) + self.S_wisdom_figurative(H_all)  # (N, |W|) + (N, |W|) -> (N, |W|)
        return S_wisdom

    def S_wisdom_figurative(self, H_all: torch.Tensor) -> torch.Tensor:
        """
        param: H_all (N, L, H)
        return: S_wisdom_figurative (N, |W|)
        """
        # 속담의 임베딩은 여기에 있습니다!
        # 참고: wisdomify/examples/explore_bert_embeddings.py
        W_embed = self.bert_mlm.bert.embeddings.word_embeddings(self.wiskeys)  # (|W|,) -> (|W|, H)

        # TODO 다음을 계산해주세요!
        S_wisdom_figurative: torch.Tensor = ...
        return S_wisdom_figurative


```

**Comment by eubinecto** [comment]
아, 생각을 해보니 `XBuilder`, `XWithWisdomMaskBuilder`, 굳이 이렇게 둘로 나눌 필요가 없을 것 같다. 그냥 기본 X포맷에 wisdom mask을 포함하면 된다.

**Comment by ohsuz** [comment]
![image](https://user-images.githubusercontent.com/59340911/136183895-7fad44e1-c806-48b1-9d72-b117c1d56cbf.png)
- BERT를 통과한 output에서 CLS Token의 임베딩과 average pooling해서 구한 속담 부분에 해당하는 [MASK] 토큰들의 임베딩, 용례 부분의 임베딩을 concat해 추가적인 linear layer에 통과시켜 문장 전체의 맥락을 담은 H_context를 구함
    - CLS token에 문장 전체 맥락 정보가 담겨있긴 하지만, [R-BERT](https://github.com/monologg/R-BERT) 논문을 참고하여 global context 이외에 local context 정보도 추가적으로 활용하고자 함
-  이렇게 구한 H_context를 속담 임베딩과 내적해 속담의 비유적인 표현을 담고자 하는... 그런....

**Comment by ohsuz** [comment]
## Result
![image](https://user-images.githubusercontent.com/59340911/136688562-492ee374-8963-4d98-a04a-247f5249da54.png)
- 이게 맞게 된건가? 이전 성능을 못 찾겠음

**Comment by eubinecto** [comment]
> 안 그래도 그 부분이 고민되긴 했는데..
> 생각해보니 eg에 해당하지 않는 부분의 히든벡터를 0으로 만드는 게 아니라 애초에 input_ids 값을 0으로 만들면(pad 토큰 취급을 한 채로 mlm을 통과시키면) 해결이 될 것 같기도 하네요! 해보겠습니다!

수지님! 그렇게 하는 것보단, 어차피  eg의 길이는 `L - (K  + 3)` (속담의 길이 K, cls, sep(중간), sep(끝) )으로 고정이니, `(N, L - (K + 3),  H)` 행렬을 출력하는 것도 괜찮을 것 같아요. 어떻게 생각하세요?

**Comment by eubinecto** [comment]
@Yuuraa  @ohsuz  V1에 대한 몇가지 궁금증 및 개선 제안입니다.


## `H_cls`를 선형변환을 해줘야만 하는 이유?
<img width="197" alt="image" src="https://user-images.githubusercontent.com/56193069/137713962-aa3c18d6-29f5-42e4-96a5-ce7ec807ce17.png">

`H_cls`를 반드시 선형변환해주어야 하는건가요? 선형변환을 해도 차원의 크기는 그대로라면, 굳이 선형변환을 해주는 명분이 무엇인가요? 
H_wisdom과  H_eg는 merge를 해야해서..라는 명분이 존재합니다. 그런데 `H_cls = linear(H_cls)`의 명분은 구조의 대칭을 위해서(?)외에는 딱히 명분을 모르겠어요. 어차피 encoder속에도 linear layer가 충분히 존재하잖아요. 그렇다면 그 내부에 있는 레이어만 미세조정하는 것으로 충분하지 않을까요?


## 왜 반드시 평균을 해줘야 하나?
왜 선형변환을 하기전에 반드시 평균을 내야만하나요? 그냥 선형변환으로 바로 차원을 축소하면 안되나요? 평균을 낸다는 것은 우리가 `H_k`속  모든 히든벡터가 equally important하다는 가정을 하는 것이잖아요. 그런데 우리는 `k`개의 마스크가 전부 속담에 대응하지 않는다는 것을 압니다 (이 부분은 RBERT랑 다른 부분이겠죠). 그럼 적어도 마지막 몇개의 히든벡터는 의미상 별 쓸모가 없어요. 그런 상황에서 평균을 해버리면 `H_wisdom`에 별 쓸모없는 정보가 들어갈 수 있다고 생각해요.

**Comment by eubinecto** [comment]
`version_2` 트레이닝이 여전히 진행이 되지 않음.  이건  config을 제대로 손을 보아야 해결할 수 있는 문제라, 일단 68은 여기서 종료.
